### PR TITLE
Use rounding in float to Duration conversion methods

### LINKED ISSUE #96045: Regression in rounding of Duration::from_secs_f64 in 1.60.0
```rust
use std::time::Duration;

fn main() {
    let dur = Duration::from_secs_f64(0.000042);
    assert_eq!(dur, Duration::new(0, 42000));
}
```

The above assertion succeeds on Rust standard library versions prior to 1.60.0. On 1.60.0 it fails with:

```console
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `41.999µs`,
 right: `42µs`', src/main.rs:6:5
```

The exact mathematical value represented by 0.000042_f64 is:

```
0.0000419999999999999976759042230600726952616241760551929473876953125
```

which is about 41999.999999999998ns. This is astronomically closer to 42000ns than to 41999ns. I believe the correct behavior for `from_secs_f64` would be to produce 42000ns as before, not truncate to 41999ns.

Mentioning @newpavlov @nagisa since #90247 seems related and is in the right commit range. The PR summary mentions that the PR would be performing truncation instead of rounding but I don't see that this decision was ever discussed by the library API team. (Mentioning @rust-lang/libs-api to comment on desired behavior.)

### PR DESCRIPTION
**Pull request by newpavlov**
Closes #96045

### CODE REVIEW

File: library/core/src/time.rs [hunk 1]
```diff
@@ -1278,13 +1276,30 @@ macro_rules! try_from_secs {
...
-            let secs = mant >> ($mant_bits - exp);
+            let secs = u64::from(mant >> ($mant_bits - exp));
             let t = <$double_ty>::from((mant << exp) & MANT_MASK);
-            let nanos = (<$double_ty>::from(NANOS_PER_SEC) * t) >> $mant_bits;
-            (u64::from(secs), nanos as u32)
+            let nanos_tmp = <$double_ty>::from(NANOS_PER_SEC) * t;
+            let nanos = (nanos_tmp >> $mant_bits) as u32;
+            if nanos_tmp & (1 << ($mant_bits - 1)) == 0 {
+                (secs, nanos)
+            } else if nanos + 1 != NANOS_PER_SEC {
```
[the remark this answers is not in the document]
Comment (newpavlov) [reply library/core/src/time.rs:1296]: Does the tie mode really matter in our case? IIUC tie to even is the default rounding in IEEE 754 only because it results in a slower rate of error accumulation for repeated additions/subtractions, which clearly is irrelevant in our case.

Implementing tie to even in software will be a bit unpleasant:
```rust
let rem_mask = (1 << nanos_offset) - 1;
let rem_msb_mask = 1 << (nanos_offset - 1);
let rem = nanos_tmp & rem_mask;
let is_tie = rem == rem_msb_mask;
let is_even = (nanos & 1) == 0;
let rem_msb = nanos_tmp & rem_msb_mask == 0;
if rem_msb || (is_even && is_tie) {
    (0, nanos)
} else {
    if nanos + 1 != NANOS_PER_SEC {
        (0, nanos + 1)
    } else {
        (1, 0)
    }
}
```

[the remark this answers is not in the document]
Comment (nagisa) [reply library/core/src/time.rs:1296]: I think it is critical that we implement the same tie resolution mechanism as used by the regular floating operations if we are implementing rounding-to-nearest for this specific operation at all.

Yes, implementing tie-away-from-zero is simpler. It is also true that the specific points raised in the referenced issue will get resolved no matter which tie resolution scheme is adopted here. I acknowledge that there are very few examples where the need to break ties comes up in the first place, and that the motivations behind picking a rounding mode for floating point operations doesn't apply here to the same degree.

What I'm worried about, however, is that somebody eventually comes up with some specific input that doesn't look “right”, would work right with a different tie resolution mode, and declares it a regression. With ties-away-from-zero their case would be that much stronger precisely because this scheme is so obscure.

Basically ties-to-even has a benefit of making it work exactly like the rest of the floating point operations, and in turn much easier to justify the specific choice for this function. If it doesn’t for some reason work right for somebody, it is then up to themselves to implement something that they think works right.

---

All that said, I'm be happy to leave the final call up to libs-api or some other T-libs team here.

**Comment by Stargateur** [comment]
there is really no tool for rounding float ? we are force to do it at hand in Duration ?

**Comment by nagisa** [comment]
> there is really no tool for rounding float ? we are force to do it at hand in Duration ?

Correctly rounded conversions to some formats such as 32-bit or 64-bit integers are provided, usually by hardware nowadays. `Duration`, however, is special in that it is a 96-bit integer with a specific format (the 32-bit nanosecond counter is expected to be in a half-open range 0..1E9) so implementing it ourselves is most likely the only reasonable approach.

**Comment by newpavlov** [comment]
Well, another theoretical alternative is to migrate `Duration` from the somewhat weird 96-bit type which does not use the 32-bit half fully and to use simple 128-bit nanosecond counter. It would require additional conversions when we talk to most OSes, but it would significantly simplify other parts.

**Comment by Stargateur** [comment]
@newpavlov to be clear I never say we should do this, I just wondering if there was a tool to manipulate float to avoid doing such complex operation in the code of `Duration`.

**Comment by nagisa** [comment]
Nominating for a T-libs decision on:

a) Do we want rounding after all?
b) What kind of tie resolution should be used if a.

**Comment by nagisa** [comment]
I don't believe there are any significant performance concerns here since we aren't able to use any hardware acceleration for the conversion either way, and I would be surprised if any ever showed up. The only concern with different tie resolution models I believe is just that one of them is just significantly more nasty to implement and understand.

File: library/core/src/time.rs [hunk 2]
```diff
@@ -1272,19 +1270,46 @@ macro_rules! try_from_secs {
...
+            let rem_msb_mask = 1 << (nanos_offset - 1);
+            let rem = nanos_tmp & rem_mask;
+            let is_tie = rem == rem_msb_mask;
+            let is_even = (nanos & 1) == 0;
+            let rem_msb = nanos_tmp & rem_msb_mask == 0;
+            let add_ns = !(rem_msb || (is_even && is_tie));
+
+            // neither `f32`, nor `f64` can represent x.999_999_999_5 exactly,
+            // so the nanos part never will be equal to 10^9
+            (secs, nanos + add_ns as u32)
```
Comment (newpavlov) [inline library/core/src/time.rs:1312]: For `f32` I have exhaustively checked that the assertion is true for all possible values, while for `f64` I tested it for a huge number of `x` starting from 1 (I left running testing loop `for x in 1.. { .. }` for a while). Unfortunately, right now I can not provide a more rigorous proof, but I am quite certain that it is true.

**Review by newpavlov — COMMENTED** [review]
@nagisa I have implemented tie to even. Can you review it?

**Comment by joshtriplett** [comment]
LGTM. @nagisa?

File: library/core/src/time.rs (hunk 2 above)
Comment (nagisa) [reply library/core/src/time.rs:1312]: Could we place a `debug_assert` to that effect here?

**Comment by nagisa** [comment]
I don't have a capacity (mental or time) at this time to rigorously recheck the implementation, but I do recall thinking for a while about the approach back when it was proposed originally in one of the comment thread, and from what I can tell it hasn't changed much since.

Given that this had two eyes on it (Josh and past me) I'm quite comfortable with a `r=me,Josh`. The only improvement that I think could be made here is that `debug_assert` though it is more of a nit, so feel free to ignore it if you don't think adding one is worthwhile. I'm going to delegate+ in case you don't have r+ powers, though I'm pretty sure you do.

@bors delegate+

File: library/core/src/time.rs (hunk 2 above)
Comment (newpavlov) [reply library/core/src/time.rs:1312]: Will do. But note that it will be nearly useless until #51713 is resolved.

Comment (newpavlov) [reply library/core/src/time.rs]: I was caring about a tie, but forgot that nanos overflow can happen without it... My check for `f32` has passed because it does not have enough precision to represent values which will trigger overflow, while the loop for `f64` was checking only for representability of x.999_999_999_5.

**Comment by inquisitivecrystal** [comment]
@newpavlov: when approving, remember to do `r=nagisa,joshtriplett` instead of `r+`.
